Execnet son!

goodboy opened the pull request:
~I need xonsh/amalgamate#7 to land before this can get in but~
I stripped out that dependency since I'm getting no feedback on that front...

This adds support for [`execnet`](http://codespeak.net/execnet/basics.html) control - an alternative (and arguably more flexible) competitor to `rpyc`.

Please feel free to rip me a new one regarding the API naming choices.

Commit:
Initial execnet support

Extend our rpc controls to include the wonderful `execnet`!
Integrate and wrap basic gateway usage such that remote
code exec can be done easily at any location.

Commit:
Move core networking routines into a utils module

Commit:
Support remote function invocation

Use `inspect.getsource()` to compile a provided function's containing
module for shipment to remote locations using `execnet`. This makes it
easy to invoke a locally defined routine remotely assuming it's
containing module can satisfy all its dependencies at the remote
location.

vodik commented on lab/ctl/rpc.py:
```diff
@@ -76,6 +79,74 @@ def get_pymod(self, modname, venvpath=None):
+                                             hostname=location.hostname)
+        keyfile = facts.get('keyfile')
+        if keyfile:
+            sshspec = "-i {} ".format(keyfile) + sshspec
```
Not a big deal, but you probably want to quote the filename.

vodik replied:
Probably still a good idea

goodboy replied:
oddly enough changing this causes me to be prompted to unlock my key?

```bash
Warning: Identity file 'id_rsa.pytest' not accessible: No such file or directory.                    
Warning: Permanently added 'sip-cannon.qa.sangoma.local,10.10.8.19' (RSA) to the list of known hosts.
Enter passphrase for key '/home/tyler/.ssh/id_rsa':
```

So i'm just going to leave as is for now and we can investigate down the road if someone wants.

vodik commented on lab/ctl/rpc.py:
```diff
@@ -76,6 +79,74 @@ def get_pymod(self, modname, venvpath=None):
+
+    def from_specdict(self, spec):
+        return execnet.makegateway('//'.join(
+            "{k}={v}".format(k=key, v=val) for key, val in spec.items()
```
```python
"=".join((key, value))
```

cleaner/shorter?

vodik commented on lab/ctl/rpc.py:
```diff
@@ -76,6 +79,74 @@ def get_pymod(self, modname, venvpath=None):
+        """
+        modpath = func.__module__
+        module = importlib.import_module(modpath)
+        pytest.set_trace()
```
Intentional 😉?

goodboy replied:
jesus...

goodboy replied:
wait no this isn't there?

Commit:
Only define some routines when `ipaddress` exists

For RPC with `execnet` we require that module dependencies are installed
at the remote location (share nothing model). For py2 it may be the case
that the `ipaddress` module is not available. For now only define
dependent network utilities if we can import the that module.

goodboy commented:
@vodik @dtkerr can I please get this reviewed asap.
Once this lands we'll have more py3 support for internal test sets.

idletea reviewed (approved):
Looks fine to me.
